src/Package/Impl/Utilities/SessionUtilities.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.R.Components.InteractiveWorkflow;
+using Microsoft.R.Host.Client;
+using Microsoft.R.Host.Client.Session;
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.R.Package.Shell;
+
+namespace Microsoft.VisualStudio.R.Package.Utilities {
+    public static class SessionUtilities {
+        public static IRSession GetInteractiveSession() {
+            var workflowProvider = VsAppShell.Current.ExportProvider.GetExportedValue<IRInteractiveWorkflowProvider>();
+            var interactiveWorkflow = workflowProvider.GetOrCreate();
+            return interactiveWorkflow.RSession;
+        }
+
+        public static async Task<string> GetRWorkingDirectoryAsync(IRInteractiveWorkflow workflow = null) {
+            var session = workflow != null ? workflow.RSession : GetInteractiveSession();
+            if (session.IsHostRunning) {
+                await TaskUtilities.SwitchToBackgroundThread();
+                try {
+                    using (var evaluation = await session.BeginEvaluationAsync(false)) {
+                        return await evaluation.GetWorkingDirectory();
+                    }
+                } catch (TaskCanceledException) { }
+            }
+            return null;
+        }
+
+        public static async Task<string> GetRUserDirectoryAsync(IRInteractiveWorkflow workflow = null) {
+            var session = workflow != null ? workflow.RSession : GetInteractiveSession();
+            if (session.IsHostRunning) {
+                await TaskUtilities.SwitchToBackgroundThread();
+                try {
+                    using (var evaluation = await session.BeginEvaluationAsync(false)) {
+                        return await evaluation.GetRUserDirectory();
+                    }
+                } catch (TaskCanceledException) { }
+            }
+            return null;
+        }
+
+        public static async Task<string> GetFriendlyDirectoryNameAsync(string directory) {
```

Comment:
This should be named `GetRShortenedPathNameAsync` or something like that, because it's also useful for files (the fix for #829 will entail calling this function when doing `source`).